## Tesla Stock

Tesla stock  price is highly manipulated by inside investors, news sentiment and products launchment. I have been reseraching for a while how to make profit on its volatile behavior.

The key is to remove the data noice, smooth out the historical price and understand how much it is really worth. 

The second important feature is to know what Tesla truly is. Is it a auto company? Is it a high tech company? Or it is mixed of traditioanl auto and high tech companies?  How similar it is relate to the traditional auto companie or the high tech companies?



In [89]:
import pandas as pd
pd.__version__

import plotly.express as px
import warnings
warnings.simplefilter(action='ignore',category=Warning)

In [59]:
import pandas_datareader as web
import datetime as dt

import matplotlib.pyplot as plt

In [45]:
tesla=pd.read_csv('TSLA.csv')
tesla=tesla[::-1]
tesla.head()

Date        Open        High         Low       Close   Adj Close  \
2377  2019-12-06  335.000000  338.859985  334.769989  335.890015  335.890015   
2376  2019-12-05  332.829987  334.420013  327.250000  330.369995  330.369995   
2375  2019-12-04  337.750000  337.859985  332.850006  333.029999  333.029999   
2374  2019-12-03  332.619995  337.910004  332.190002  336.200012  336.200012   
2373  2019-12-02  329.399994  336.380005  328.690002  334.869995  334.869995   

       Volume  
2377  7618937  
2376  3724600  
2375  5533000  
2374  6573700  
2373  6074500

In [111]:
rolling_back_days=[7,14,30,60,120]

tesla['7days_rolling_back_avg']=tesla['Adj Close'].rolling(7).mean().shift(-6)
tesla['14days_rolling_back_avg']=tesla['Adj Close'].rolling(14).mean().shift(-13)
tesla['30days_rolling_back_avg']=tesla['Adj Close'].rolling(30).mean().shift(-29)
tesla['60days_rolling_back_avg']=tesla['Adj Close'].rolling(60).mean().shift(-59)
tesla['120days_rolling_back_avg']=tesla['Adj Close'].rolling(120).mean().shift(-119)
tesla['150days_rolling_back_avg']=tesla['Adj Close'].rolling(150).mean().shift(-149)
tesla['365days_rolling_back_avg']=tesla['Adj Close'].rolling(365).mean().shift(-364)
tesla['720days_rolling_back_avg']=tesla['Adj Close'].rolling(720).mean().shift(-719)


In [112]:
tesla.head()

Date        Open        High         Low       Close   Adj Close  \
2377  2019-12-06  335.000000  338.859985  334.769989  335.890015  335.890015   
2376  2019-12-05  332.829987  334.420013  327.250000  330.369995  330.369995   
2375  2019-12-04  337.750000  337.859985  332.850006  333.029999  333.029999   
2374  2019-12-03  332.619995  337.910004  332.190002  336.200012  336.200012   
2373  2019-12-02  329.399994  336.380005  328.690002  334.869995  334.869995   

       Volume  7days_rolling_back_avg  14days_rolling_back_avg  \
2377  7618937              333.084290               339.032144   
2376  3724600              332.088575               340.195001   
2375  5533000              332.941432               341.550716   
2374  6573700              332.942862               342.485000   
2373  6074500              335.604287               343.465713   

      30days_rolling_back_avg  60days_rolling_back_avg  \
2377               334.611668               290.745501   
2376               333.404667               289.245167   
2375               330.881667               287.857334   
2374               328.300000               286.232500   
2373               325.543333               284.492333   

      120days_rolling_back_avg  150days_rolling_back_avg  \
2377                261.410833                251.413400   
2376                260.484583                250.876400   
2375                259.606750                250.374133   
2374                258.622500                249.781267   
2373                257.603417                249.100000   

      365days_rolling_back_avg  720days_rolling_back_avg  
2377                284.151479                301.281542  
2376                284.168219                301.166347  
2375                284.175452                301.058209  
2374                284.177096                300.949097  
2373                284.208082                300.835778

In [131]:
columns=['Date','Adj Close','7days_rolling_back_avg','14days_rolling_back_avg',\
         '30days_rolling_back_avg','60days_rolling_back_avg','120days_rolling_back_avg','150days_rolling_back_avg','365days_rolling_back_avg','720days_rolling_back_avg']

In [142]:
tesla_df=tesla[columns]
tesla_df['Date']=pd.to_datetime(tesla_df['Date'])
tesla_df.head(20)

Date   Adj Close  7days_rolling_back_avg  14days_rolling_back_avg  \
2377 2019-12-06  335.890015              333.084290               339.032144   
2376 2019-12-05  330.369995              332.088575               340.195001   
2375 2019-12-04  333.029999              332.941432               341.550716   
2374 2019-12-03  336.200012              332.942862               342.485000   
2373 2019-12-02  334.869995              335.604287               343.465713   
2372 2019-11-29  329.940002              338.082860               344.195713   
2371 2019-11-27  331.290009              342.308572               344.710000   
2370 2019-11-26  328.920013              344.979998               345.013572   
2369 2019-11-25  336.339996              348.301426               344.846427   
2368 2019-11-22  333.040009              350.159999               343.480713   
2367 2019-11-21  354.829987              352.027139               342.368569   
2366 2019-11-20  352.220001              351.327140               339.402856   
2365 2019-11-19  359.519989              350.308567               336.738571   
2364 2019-11-18  349.989990              347.111428               333.559287   
2363 2019-11-15  352.170013              345.047145               331.147145   
2362 2019-11-14  349.350006              341.391427               329.400000   
2361 2019-11-13  346.109985              336.801427               327.884286   
2360 2019-11-12  349.929993              332.710000               324.567858   
2359 2019-11-11  345.089996              327.478572               317.764287   
2358 2019-11-08  337.140015              323.168575               311.370716   

      30days_rolling_back_avg  60days_rolling_back_avg  \
2377               334.611668               290.745501   
2376               333.404667               289.245167   
2375               330.881667               287.857334   
2374               328.300000               286.232500   
2373               325.543333               284.492333   
2372               322.946001               282.702000   
2371               320.680334               281.029333   
2370               318.295667               279.185833   
2369               315.928000               277.454000   
2368               313.282000               275.608500   
2367               310.443667               273.753000   
2366               306.774000               271.432333   
2365               303.184334               269.130000   
2364               299.202001               266.721333   
2363               295.459668               264.411500   
2362               291.435001               262.244500   
2361               287.557667               260.102500   
2360               284.125001               258.098333   
2359               280.617001               256.046667   
2358               277.143001               253.960834   

      120days_rolling_back_avg  150days_rolling_back_avg  \
2377                261.410833                251.413400   
2376                260.484583                250.876400   
2375                259.606750                250.374133   
2374                258.622500                249.781267   
2373                257.603417                249.100000   
2372                256.556667                248.458800   
2371                255.616333                247.869000   
2370                254.629583                247.228000   
2369                253.592750                246.686067   
2368                252.506166                246.168200   
2367                251.369083                245.707266   
2366                250.025500                245.093400   
2365                248.581750                244.567000   
2364                247.128750                243.978400   
2363                245.780667                243.467533   
2362                244.428083                242.895600   
2361                243.089333                242.351267   
2360                241.793667 

In [141]:
tesla_df_1=tesla_df.melt(id_vars=['Date']+list(tesla_df.keys()[2:4]),var_name=['Tesla Stock Price'])
fig=px.line(tesla_df_1,x='Date',y='value',color='Tesla Stock Price')
fig.show()

In [139]:
tesla_df_1=tesla_df.melt(id_vars=['Date']+list(tesla_df.keys()[1:8]),var_name=['Tesla Stock Price'])
fig=px.line(tesla_df_1,x='Date',y='value',color='Tesla Stock Price')
fig.show()

### Step 2

Run xgboost on sentimental analysis to see if it is worth to short as its price is over $330 today, if my time allows

### Step 3

Run similarity analysis on Ford, Nvidia and Tesla, if my time allows